From 8fd087c98dd5887638537749fa68ce6696e0d07a Mon Sep 17 00:00:00 2001
From: Alex Turbov <i.zaufi@gmail.com>
Date: Tue, 16 Jun 2015 01:16:57 +0700
Subject: [PATCH] Fix the module to handle SYSCONFDIR and LOCALSTATEDIR
 properly if CMAKE_INSTALL_PREFIX is set to `/` or `/usr` -- i.e. as expected
 by GNU Coding Standard (i.e. set SYSCONFDIR to `/etc` and `LOCALSTATEDIR` to
 `/var`).

Also if CMAKE_INSTALL_PREFIX is set to /opt/pkg, `SYSCONFDIR` must be set to
`/etc/opt/pkg` and `LOCALSTATEDIR` to `/var/opt/pkg` according to FHS.
---
 Modules/GNUInstallDirs.cmake | 48 ++++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 46 insertions(+), 2 deletions(-)

diff --git a/Modules/GNUInstallDirs.cmake b/Modules/GNUInstallDirs.cmake
index c61e7e9..14055d5 100644
--- a/Modules/GNUInstallDirs.cmake
+++ b/Modules/GNUInstallDirs.cmake
@@ -55,8 +55,27 @@
 # CMAKE_INSTALL_FULL_<dir> value contains an absolute path constructed
 # from the corresponding destination by prepending (if necessary) the
 # value of CMAKE_INSTALL_PREFIX.
+#
+# There are few special install prefixes are handled in a special way.
+# According to GNU coding style:
+#
+# > When building the complete GNU system, the prefix will be empty and ``/usr`` will be a symbolic link to ``/``.
+#
+# I.e. if someone wants to install a package to ``/`` (like RPM/DEB/etc packages do),
+# directories that are not a part of ``/usr/`` will stay at the root of the filesystem
+# (e.g. ``SYSCONFDIR`` and/or ``LOCALSTATEDIR``).
+#
+# The second part of the problem: there are few people who know about this GNU feature,
+# so specifying ``-DCMAKE_INSTALL_PREFIX=/usr`` should act the same way.
+#
+# Another case is when the user wants to install his package to ``/opt/pkg``. According to
+# Filesystem Hierarchy Standard (https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html),
+# ``SYSCONFDIR`` should be ``/etc/opt/pkg`` and ``LOCALSTATEDIR`` is ``/var/opt/pkg``.
+#
+
 
 #=============================================================================
+# Copyright 2015 Alex Turbov <i.zaufi@gmail.com>
 # Copyright 2011 Nikita Krupen'ko <krnekit@gmail.com>
 # Copyright 2011 Kitware, Inc.
 #
@@ -274,8 +293,33 @@ foreach(dir
     MANDIR
     DOCDIR
     )
-  if(NOT IS_ABSOLUTE ${CMAKE_INSTALL_${dir}})
-    set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_${dir}}")
+  if(NOT IS_ABSOLUTE "${CMAKE_INSTALL_${dir}}")
+    # Handle special cases:
+    # - CMAKE_INSTALL_PREFIX == /
+    # - CMAKE_INSTALL_PREFIX == /usr
+    if("${CMAKE_INSTALL_PREFIX}" STREQUAL "/")
+        if("${dir}" STREQUAL "SYSCONFDIR" OR "${dir}" STREQUAL "LOCALSTATEDIR")
+            set(CMAKE_INSTALL_FULL_${dir} "/${CMAKE_INSTALL_${dir}}")
+        else()
+            set(CMAKE_INSTALL_${dir} "usr/${CMAKE_INSTALL_${dir}}")
+            set(CMAKE_INSTALL_FULL_${dir} "/${CMAKE_INSTALL_${dir}}")
+        endif()
+    elseif("${CMAKE_INSTALL_PREFIX}" MATCHES "^/usr/?$")
+        if("${dir}" STREQUAL "SYSCONFDIR" OR "${dir}" STREQUAL "LOCALSTATEDIR")
+            set(CMAKE_INSTALL_FULL_${dir} "/${CMAKE_INSTALL_${dir}}")
+        else()
+            set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_${dir}}")
+        endif()
+    # - CMAKE_INSTALL_PREFIX == /opt
+    elseif("${CMAKE_INSTALL_PREFIX}" MATCHES "^/opt/.*")
+        if("${dir}" STREQUAL "SYSCONFDIR" OR "${dir}" STREQUAL "LOCALSTATEDIR")
+            set(CMAKE_INSTALL_FULL_${dir} "/${CMAKE_INSTALL_${dir}}${CMAKE_INSTALL_PREFIX}")
+        else()
+            set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_${dir}}")
+        endif()
+    else()
+        set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_${dir}}")
+    endif()
   else()
     set(CMAKE_INSTALL_FULL_${dir} "${CMAKE_INSTALL_${dir}}")
   endif()
-- 
2.4.3

